File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        res = []
        if not root:
            return res
        queue = [[root]]

        while queue:
            print("New level")
            curr_level = queue[0]
            queue.pop(0)

            now = []
            res_curr = []
            for i in curr_level:
                # print(i.val)
                res_curr.append(i.val)


                
                if i.left: now.append(i.left)
                if i.right: now.append(i.right)

            if len(now) > 0:
                queue.append(now) 

            res.append(res_curr)
        return res
            

File: Data Structures & Algorithms/balanced-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool res = true;
    int traverse(TreeNode* node, int depth){
        if(node == NULL) return 0;
        int left = traverse(node->left, depth+1);
        std::cout << node->val << " " << depth << std::endl;
        int right = traverse(node->right, depth+1);
        std::cout << "Max:" << max(left, right) << std::endl;
        std::cout << "Max diff:" << left - right << std::endl;
        if(abs(left-right)>1) res = false;
        return max(left, right) + 1;
    }
    bool isBalanced(TreeNode* root) {
        traverse(root, 0);
        return res;
    }
};
